File: LC169_MajorityElement/Hash/hash.py
# Given an array nums of size n, return the majority element.
# The majority element is the element that appears more than ⌊n / 2⌋ times. You may assume that the majority element always exists in the array.
from typing import List

class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        # Hash
        # Time Complexity: O(n)
        # Space Complexity: O(n)
        mapping = {}
        for num in nums:
            if num not in mapping:
                mapping[num] = 0
            mapping[num] = mapping[num] + 1
        half_length = len(nums) // 2
        for key in mapping:
            if mapping[key] > half_length:
                return key

File: LC13_RomanToInteger/Dictionary/hash.cpp
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    int romanToInt(string s) {
        // Iteration
        // Time Complexity: O(1)
        // Space Complexity: O(1)
        unordered_map<char, int> mapping {
            {'I', 1}, {'V', 5}, {'X', 10}, {'L', 50}, {'C', 100}, {'D', 500}, {'M', 1000}
            };
        if(s.length() == 0){
            return 0;
        }
        int result = mapping[s[0]];
        for(int i = 1; i < s.length(); i++){
            if(mapping[s[i]] > mapping[s[i-1]]){
                result = result + mapping[s[i]] - 2*mapping[s[i-1]];
            }else{
                result += mapping[s[i]];
            }
        }
        return result;
    }
};